### PR TITLE
feat: Trivy Image Scanning

### PR DESCRIPTION
**Pull request by Shivam-nagar23**
# Description

Currently we support Vulnerability Image scanning through clair, we are integrating the support of Trivy Image Vulnerabilities  Scanning.
Fixes [#AB2842](https://dev.azure.com/DevtronLabs/Devtron/_workitems/edit/2842)
FE-PR -[FE-PR](https://github.com/devtron-labs/dashboard/pull/982)
Image scanner- [PR](https://github.com/devtron-labs/image-scanner/pull/15)
## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [x] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: api/module/ModuleRestHandler.go [hunk 1]
```diff
@@ -164,3 +165,46 @@ func (impl ModuleRestHandlerImpl) HandleModuleAction(w http.ResponseWriter, r *h
...
+	moduleName := params["name"]
+	if len(moduleName) == 0 {
+		impl.logger.Error("module name is not supplied")
+		common.WriteJsonResp(w, errors.New("module name is not supplied"), nil, http.StatusBadRequest)
+		return
+	}
+	// decode request
+	decoder := json.NewDecoder(r.Body)
+	var moduleEnableRequestDto *module.ModuleEnableRequestDto
+	err = decoder.Decode(&moduleEnableRequestDto)
```
Comment (kartik-579) [inline api/module/ModuleRestHandler.go:188]: dont use pointer of pointer

Comment (kartik-579) [reply api/module/ModuleRestHandler.go:188]: initialize non-pointer type

File: api/module/ModuleRestHandler.go [hunk 2]
```diff
@@ -164,3 +165,46 @@ func (impl ModuleRestHandlerImpl) HandleModuleAction(w http.ResponseWriter, r *h
...
+	token := r.Header.Get("token")
+	if ok := impl.enforcer.Enforce(token, casbin.ResourceGlobal, casbin.ActionUpdate, "*"); !ok {
+		common.WriteJsonResp(w, errors.New("unauthorized"), nil, http.StatusForbidden)
+		return
+	}
+
+	// service call
+	res, err := impl.moduleService.EnableModule(moduleName, moduleEnableRequestDto.Version)
+	if err != nil {
+		impl.logger.Errorw("service err, Enabling Module", "err", err)
```
Comment (kartik-579) [inline api/module/ModuleRestHandler.go:205]: add metadata in logs

File: internal/sql/repository/security/ScanToolExecutionHistoryMapping.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	return nil
+}
+
+func (repo *ScanToolExecutionHistoryMappingRepositoryImpl) UpdateStateByToolAndExecutionHistoryId(executionHistoryId, toolId int,
+	state serverBean.ScanExecutionProcessState, executionFinishTime time.Time) error {
+	model := &ScanToolExecutionHistoryMapping{}
+	_, err := repo.dbConnection.Model(model).Set("state = ?", state).
+		Set("execution_finish_time  = ?", executionFinishTime).
+		Where("image_scan_execution_history_id = ?", executionHistoryId).
+		Where("scan_tool_id = ?", toolId).Update()
```
Comment (kartik-579) [inline internal/sql/repository/security/ScanToolExecutionHistoryMapping.go:70]: update auditLog too (updated_on, updated_by)

File: internal/sql/repository/security/ScanToolExecutionHistoryMapping.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		repo.logger.Errorw("error in ScanToolExecutionHistoryMappingRepository, SaveInBatch", "err", err, "model", model)
+		return err
+	}
+	return nil
+}
+
+func (repo *ScanToolExecutionHistoryMappingRepositoryImpl) MarkAllRunningStateAsFailedHavingTryCountReachedLimit(tryCount int) error {
+	var models []*ScanToolExecutionHistoryMapping
+	_, err := repo.dbConnection.Model(&models).
+		Set("state = ?", serverBean.ScanExecutionProcessStateFailed).
```
Comment (kartik-579) [inline internal/sql/repository/security/ScanToolExecutionHistoryMapping.go:81]: update auditLog too (updated_on, updated_by)

File: pkg/module/ModuleService.go [hunk 5]
```diff
@@ -291,15 +326,54 @@ func (impl ModuleServiceImpl) HandleModuleAction(userId int32, moduleName string
...
 	module.Version = moduleActionRequest.Version
 	module.UpdatedOn = time.Now()
+	tx, err := impl.moduleRepository.GetConnection().Begin()
+	if err != nil {
+		impl.logger.Errorw("error in  opening an transaction", "err", err)
+		return nil, err
+	}
+	defer tx.Rollback()
+	res := strings.Split(moduleName, ".")
+	toolName := strings.ToUpper(res[1])
```
Comment (kartik-579) [inline pkg/module/ModuleService.go:336]: what if res has only one element, we might get index out of bound error

File: pkg/module/ModuleService.go [hunk 6]
```diff
@@ -291,15 +326,54 @@ func (impl ModuleServiceImpl) HandleModuleAction(userId int32, moduleName string
...
+			if err == pg.ErrNoRows {
+				var toolversion string
+				if moduleName == ModuleNameSecurityClair {
+					toolversion = CLAIR_V4
+				} else if moduleName == ModuleNameSecurityTrivy {
+					toolversion = TRIVY_V1
+				}
+				err2 := impl.scanToolMetaDataRepository.MarkToolAsActive(toolName, toolversion, tx)
+				if err2 != nil {
+					impl.logger.Errorw("error in marking tool as active ", "err", err)
```
Comment (kartik-579) [inline pkg/module/ModuleService.go:353]: add metadata in logs

File: pkg/module/ModuleService.go [hunk 7]
```diff
@@ -343,22 +417,77 @@ func (impl ModuleServiceImpl) HandleModuleAction(userId int32, moduleName string
...
+	res := strings.Split(moduleName, ".")
+	toolName := strings.ToUpper(res[1])
+	err = impl.moduleRepository.MarkModuleAsEnabledWithTransaction(moduleName, tx)
+	if err != nil {
+		impl.logger.Errorw("error in updating module as active ", "moduleName", moduleName, "err", err)
+		return nil, err
+	}
+	err = impl.scanToolMetaDataRepository.MarkToolAsActive(toolName, version, tx)
+	if err != nil {
+		impl.logger.Errorw("error in marking tool as active ", "err", err)
```
Comment (kartik-579) [inline pkg/module/ModuleService.go:457]: add metadata in log

File: pkg/module/ModuleService.go [hunk 8]
```diff
@@ -343,22 +417,77 @@ func (impl ModuleServiceImpl) HandleModuleAction(userId int32, moduleName string
...
+		return nil, err
+	}
+	err = impl.scanToolMetaDataRepository.MarkOtherToolsInActive(toolName, tx, version)
+	if err != nil {
+		impl.logger.Errorw("error in marking other tools inactive ", "err", err)
```
Comment (kartik-579) [inline pkg/module/ModuleService.go:462]: add metadata in log

File: pkg/module/ModuleService.go [hunk 9]
```diff
@@ -343,22 +417,77 @@ func (impl ModuleServiceImpl) HandleModuleAction(userId int32, moduleName string
...
+		return nil, err
+	}
+	err = impl.moduleRepository.MarkOtherModulesDisabled(moduleName, module.ModuleType, tx)
+	if err != nil {
+		impl.logger.Errorw("error in marking other modules of same module type inactive ", "err", err)
```
Comment (kartik-579) [inline pkg/module/ModuleService.go:467]: add metadata in log

File: pkg/module/ModuleService.go [hunk 10]
```diff
@@ -384,9 +513,21 @@ func (impl ModuleServiceImpl) GetAllModuleInfo() ([]ModuleInfoDto, error) {
...
+			Moduletype: module.ModuleType,
+			Enabled:    module.Enabled,
+		}
+		enabled := false
+		if len(module.ModuleType) == 0 {
+			module.Enabled = true
+			enabled = true
+			err := impl.moduleRepository.Update(&module)
+			if err != nil {
+				impl.logger.Errorw("error in updating installed module to rnabled for previous modules")
```
Comment (kartik-579) [inline pkg/module/ModuleService.go:527]: add Error & metadata in log. why are we not returning from here?

File: pkg/security/ImageScanService.go [hunk 11]
```diff
@@ -373,6 +378,17 @@ func (impl ImageScanServiceImpl) FetchExecutionDetailResult(request *ImageScanRe
 			}
 			executionTime = item.ImageScanExecutionHistory.ExecutionTime
 		}
+		if len(imageScanResult) > 0 {
+			imageScanResponse.ScanToolId = imageScanResult[0].ScanToolId
+		} else {
+			scanToolHistoryMappings, err := impl.scanToolExecutionHistoryRepository.GetAllScanHistoriesByExecutionHistoryIds(scanExecutionIds)
+			if err != nil {
```
Comment (kartik-579) [inline pkg/security/ImageScanService.go:385]: why are we not returning from here? we might get connection error or something else. If empty result is okay, then add check for NoRows and return for remaining errors

File: pkg/security/ImageScanService.go [hunk 12]
```diff
@@ -479,6 +495,17 @@ func (impl ImageScanServiceImpl) FetchMinScanResultByAppIdAndEnvId(request *Imag
 				lowCount = lowCount + 1
 			}
 		}
+		if len(imageScanResult) > 0 {
+			scantoolId = imageScanResult[0].ScanToolId
+		} else {
+			scanToolHistoryMappings, err := impl.scanToolExecutionHistoryRepository.GetAllScanHistoriesByExecutionHistoryIds(scanExecutionIds)
+			if err != nil {
+				impl.Logger.Errorw("error in getting scanToolHistoryMappings", "err", err)
```
Comment (kartik-579) [inline pkg/security/ImageScanService.go:503]: same comment as above. We can extract the logic (something like getScanToolIdForReasonX) to remove duplicacy

File: scripts/sql/145_generic_image_scanning.up.sql [hunk 13]
```diff
@@ -0,0 +1,121 @@
+CREATE SEQUENCE IF NOT EXISTS id_seq_scan_tool_metadata;
+
+CREATE TABLE public.scan_tool_metadata
+(
+    "id"                                  integer NOT NULL DEFAULT nextval('id_seq_scan_tool_metadata'::regclass),
+    "name"                                text,
```
Comment (kartik-579) [inline scripts/sql/145_generic_image_scanning.up.sql:6]: change to varchar(100)

File: scripts/sql/145_generic_image_scanning.up.sql [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    "version"                             VARCHAR(50),
+    "server_base_url"                     text,
+    "result_descriptor_template"            text,
+    "scan_target"                          text,
```
Comment (kartik-579) [inline scripts/sql/145_generic_image_scanning.up.sql:10]: change to varchar

File: scripts/sql/145_generic_image_scanning.up.sql [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    CONSTRAINT "scan_tool_execution_history_mapping_result_id_fkey" FOREIGN KEY ("image_scan_execution_history_id") REFERENCES "public"."image_scan_execution_history" ("id"),
+    CONSTRAINT "scan_tool_execution_history_mapping_scan_tool_id_fkey" FOREIGN KEY ("scan_tool_id") REFERENCES "public"."scan_tool_metadata" ("id")
+);
+
+CREATE SEQUENCE IF NOT EXISTS id_seq_scan_step_condition;
+
+CREATE TABLE public.scan_step_condition
+(
+    "id"                   integer NOT NULL DEFAULT nextval('id_seq_scan_step_condition'::regclass),
+    "condition_variable_format" text,
```
Comment (kartik-579) [inline scripts/sql/145_generic_image_scanning.up.sql:46]: varchar

**Review by kartik-579 — CHANGES_REQUESTED** [review]

File: pkg/app/AppService.go [hunk 16]
```diff
@@ -2446,6 +2446,12 @@ func (impl *AppServiceImpl) MarkImageScanDeployed(appId int, envId int, imageDig
 			impl.logger.Errorw("error in creating deploy info", "err", err)
 		}
 	} else {
+		// Updating Execution history for Latest Deployment to fetch out security Vulnerabilities for latest deployed info
+		ot.ImageScanExecutionHistoryId = ids
+		err := impl.imageScanDeployInfoRepository.Update(ot)
```
Comment (kripanshdevtron) [inline pkg/app/AppService.go:2451]: this err is not returned as it's scope is this block only.

File: pkg/app/AppService.go [hunk 17]
```diff
@@ -2446,6 +2446,12 @@ func (impl *AppServiceImpl) MarkImageScanDeployed(appId int, envId int, imageDig
...
+		if err != nil {
+			impl.logger.Errorw("error in updating deploy info for latest deployed image", "err", err)
+		}
 		impl.logger.Debugw("pt", "ot", ot)
```
[the remark this answers is not in the document]
Comment (Shivam-nagar23) [reply pkg/app/AppService.go:2455]: removed the log

File: pkg/module/ModuleService.go (hunk 6 above)
Comment (kripanshdevtron) [reply pkg/module/ModuleService.go:357]: why are we logging err here instead of err2 ??

File: pkg/module/ModuleService.go [hunk 18]
```diff
@@ -291,15 +327,59 @@ func (impl ModuleServiceImpl) HandleModuleAction(userId int32, moduleName string
...
+				if moduleName == ModuleNameSecurityClair {
+					// Handled for V4 for CLAIR as we are not using CLAIR V2 anymore.
+					toolversion = CLAIR_V4
+				} else if moduleName == ModuleNameSecurityTrivy {
+					toolversion = TRIVY_V1
+				}
+				err2 := impl.scanToolMetaDataRepository.MarkToolAsActive(toolName, toolversion, tx)
+				if err2 != nil {
+					impl.logger.Errorw("error in marking tool as active ", "err", err)
+					return nil, err
```
Comment (kripanshdevtron) [inline pkg/module/ModuleService.go:358]: why returning err instead of err2

File: pkg/module/ModuleService.go [hunk 19]
```diff
@@ -343,22 +423,84 @@ func (impl ModuleServiceImpl) HandleModuleAction(userId int32, moduleName string
...
+	tx, err := dbConnection.Begin()
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+	res := strings.Split(moduleName, ".")
+	// Handling for future tools if integrated
+	if len(res) < 2 {
+		impl.logger.Errorw("error in getting toolName from modulename as module Length is smaller than 2")
+		return nil, err
```
Comment (kripanshdevtron) [inline pkg/module/ModuleService.go:458]: err would be nil here, you should use errors.New here to throw custom error

File: scripts/sql/145_generic_image_scanning.up.sql [hunk 20]
```diff
@@ -0,0 +1,121 @@
+CREATE SEQUENCE IF NOT EXISTS id_seq_scan_tool_metadata;
+
+CREATE TABLE public.scan_tool_metadata
+(
+    "id"                                  integer NOT NULL DEFAULT nextval('id_seq_scan_tool_metadata'::regclass),
+    "name"                                VARCHAR(100),
+    "version"                             VARCHAR(50),
+    "server_base_url"                     text,
```
Comment (kripanshdevtron) [inline scripts/sql/145_generic_image_scanning.up.sql:8]: why is serverBaseUrl text type ??

File: scripts/sql/145_generic_image_scanning.up.sql [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+    "scan_tool_id"   integer,
+    "index"                integer,
+    "step_execution_type"  VARCHAR(10),
+    "step_execution_sync"   bool NOT NULL ,
+    "retry_count"          integer,
+    "execute_step_on_fail" integer,
+    "execute_step_on_pass" integer,
+    "render_input_data_from_step" integer,
+    "http_input_payload"        jsonb,
+    "http_method_type"     text,
```
Comment (kripanshdevtron) [inline scripts/sql/145_generic_image_scanning.up.sql:72]: why is this text type ?? ``text type are slows``

**Review by kripanshdevtron — APPROVED** [review]

**Review by kartik-579 — APPROVED** [review]

**Review by kartik-579 — APPROVED** [review]

**Comment by prakarsh-dt** [comment]
Duplicate of #1338

**Comment by prakarsh-dt** [comment]
Fixes #1338

**Pull request merged**